### PR TITLE
GPS handler

### PR DESCRIPTION
**Pull request by hitimo**
This pull request is only for commenting on the code.

### CODE REVIEW

File: gps_rviz/src/gps_handler.cpp [hunk 1]
```diff
@@ -0,0 +1,122 @@
+#include "geometry_msgs/Vector3.h"
+#include "math.h"
+#include "visualization_msgs/Marker.h"
+#include "visualization_msgs/MarkerArray.h"
+#include "gps_handler.h"
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: We use this include style: https://github.com/ethz-asl/multiagent_mapping/wiki/Header-include-guide
Alphabetize includes.

File: gps_rviz/src/gps_handler.cpp [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+#define PI 3.14159265
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. Use M_PI in the code. Also avoid using macros.

File: gps_rviz/src/gps_handler.cpp [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+using namespace std;
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. Don't use `using namespace ...` but e.g. `std::cout` in the code.

File: gps_rviz/src/gps_handler.cpp [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+namespace gps{
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. hspace: `gps {`

File: gps_rviz/src/gps_handler.cpp [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+ geometry_msgs::Vector3 InitializingPoint, InitializingPoint_missionframe;
+ bool REFERENCE_FRAME_IS_INITIALIZED;
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE.
If variable is constant:
`bool kIsReferenceFrameInitialized = false;`
or if variable may change:
`bool is_reference_frame_initialized;`

File: gps_rviz/src/gps_handler.cpp [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+#include "visualization_msgs/MarkerArray.h"
+#include "gps_handler.h"
+
+#define PI 3.14159265
+
+using namespace std;
+
+namespace gps{
+
+ geometry_msgs::Vector3 InitializingPoint, InitializingPoint_missionframe;
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: ``geometry_msgs::Vector3 initializing_point;``
``geometry_msgs::Vector3 initializing_point_mission_frame;``

File: gps_rviz/src/gps_handler.cpp [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+ bool REFERENCE_FRAME_IS_INITIALIZED;
+ float alpha_init, beta_init;
+ 
+ visualization_msgs::MarkerArray GPS_TRACK;
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. `gps_track`

File: gps_rviz/src/gps_handler.cpp [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+ visualization_msgs::Marker gps_points, gps_line_strip, gps_line_list;
+
+// ---------------------------------------------
+// Rotational Matrix R_0G=R_01*R_1G; R_01=Rz(a); R_1G=Ry(b)
+geometry_msgs::Vector3 transformation(geometry_msgs::Vector3 Input, float a, float b){
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. `input` (lower-case)

Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. hspace before `{`

File: gps_rviz/src/gps_handler.cpp [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+	geometry_msgs::Vector3 Output;
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. `output`

File: gps_rviz/src/gps_handler.cpp [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+ 	Output.x = cos(a)*cos(b)*Input.x - sin(a)*Input.y +cos(a)*sin(b)*Input.z;
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: hspace between `*`, `+` etc.

File: gps_rviz/src/gps_handler.cpp [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+ 	Output.y = sin(a)*cos(b)*Input.x + cos(a)*Input.y +sin(a)*sin(b)*Input.z;
+ 	Output.z = -sin(b)*Input.x + cos(b)*Input.z;
+
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp:22]: remove vspace.

File: gps_rviz/src/gps_handler.cpp (hunk 9 above)
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. We use 2-pt indentation.

File: gps_rviz/src/gps_handler.cpp [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+return Output;
+}
+
+// -------------------------------------------------
+// Rotational Matrix plus shifting
+geometry_msgs::Vector3 transformToMissionFrame(geometry_msgs::Vector3 InputPoint){
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. hspace before `{`

Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. `input_point`

File: gps_rviz/src/gps_handler.cpp [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+	geometry_msgs::Vector3 Output;
+ 	Output.x = cos(a)*cos(b)*Input.x - sin(a)*Input.y +cos(a)*sin(b)*Input.z;
+ 	Output.y = sin(a)*cos(b)*Input.x + cos(a)*Input.y +sin(a)*sin(b)*Input.z;
+ 	Output.z = -sin(b)*Input.x + cos(b)*Input.z;
+
+return Output;
+}
+
+// -------------------------------------------------
+// Rotational Matrix plus shifting
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp:26]: DONE. Sentensify comments.

File: gps_rviz/src/gps_handler.cpp [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+geometry_msgs::Vector3 transformToMissionFrame(geometry_msgs::Vector3 InputPoint){
+
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. Remove vspace.

File: gps_rviz/src/gps_handler.cpp [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+	geometry_msgs::Vector3 transformed_point;
+
+	transformed_point = transformation(InputPoint, alpha_init, beta_init); //Rotate by alpha, beta
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE.
Write comments above line and sentensify:

```
// Rotate by alpha and beta.
transformed_point = transformation(input_point, alpha_init, beta_init);
```

File: gps_rviz/src/gps_handler.cpp [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+ 	Output.z = -sin(b)*Input.x + cos(b)*Input.z;
+
+return Output;
+}
+
+// -------------------------------------------------
+// Rotational Matrix plus shifting
+geometry_msgs::Vector3 transformToMissionFrame(geometry_msgs::Vector3 InputPoint){
+
+	geometry_msgs::Vector3 transformed_point;
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. We use 2pt-indentation.

File: gps_rviz/src/gps_handler.cpp (hunk 12 above)
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. 
- Return value as function argument.
- Pass input arguments that do not change as pointers or by const reference.

```
void transformToMissionFrame(const geometry_msgs::Vector3& input_point, geometry_msgs::Vector3& transformed_point) {
```

File: gps_rviz/src/gps_handler.cpp [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+
+	transformed_point = transformation(InputPoint, alpha_init, beta_init); //Rotate by alpha, beta
+	transformed_point.x = transformed_point.x - InitializingPoint_missionframe.x; // shift to normalize
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. 

```
// Shift to normalize.
transformed_point.x = transformed_point.x - initializing_point_mission_frame.x; 
```

File: gps_rviz/src/gps_handler.cpp [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+	transformed_point.y = transformed_point.y - InitializingPoint_missionframe.y;
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. same

File: gps_rviz/src/gps_handler.cpp [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+	transformed_point.z = transformed_point.z - InitializingPoint_missionframe.z;
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. same

File: gps_rviz/src/gps_handler.cpp [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. Remove vspace.

File: gps_rviz/src/gps_handler.cpp [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+return transformed_point;
+} 
+
+// ---------------------------------------------------
+// NavSatFix coordinates to WGS84 (cartesian) coordinates
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. Sentensify comment.

File: gps_rviz/src/gps_handler.cpp [hunk 22]
```diff
@@ -0,0 +1,122 @@
...
+geometry_msgs::Vector3 gps2cart(const geometry_msgs::Vector3& gpsOut){
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. 
- Avoid abbreviations in names. 
- Better name for `gps_out`?

```
void gpsToCartesian(const geometry_msgs::Vector3& gps_out, geometry_msgs::Vector3& gps_cartesian) { 
```

File: gps_rviz/src/gps_handler.cpp [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. Remove vspace.

File: gps_rviz/src/gps_handler.cpp [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+return transformed_point;
+} 
+
+// ---------------------------------------------------
+// NavSatFix coordinates to WGS84 (cartesian) coordinates
+geometry_msgs::Vector3 gps2cart(const geometry_msgs::Vector3& gpsOut){
+	
+	float e, R, height, phi_deg, phi_rad, lambda_deg, lambda_rad; 
+
+	R = (majorAxis+minorAxis)/2; //in meters
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. 
- Use unit in variable name and remove comment.
- Longer variable names.
- Code should be self-explanatory.
- Take care of double precision when using `*` and `/`.

```
r_meter = (major_axis + minor_axis) / 2.0;
```

File: gps_rviz/src/gps_handler.cpp [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+	e = (majorAxis*majorAxis-minorAxis*minorAxis)/(majorAxis*majorAxis); //in meters
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. same

File: gps_rviz/src/gps_handler.cpp [hunk 26]
```diff
@@ -0,0 +1,122 @@
...
+
+	height 		= gpsOut.z; 		//meters
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. `height_meter = gps_out.z;`

File: gps_rviz/src/gps_handler.cpp [hunk 27]
```diff
@@ -0,0 +1,122 @@
...
+	phi_deg		= gpsOut.x; 		//degrees
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. `phi_deg = gps_out.x;`

File: gps_rviz/src/gps_handler.cpp [hunk 28]
```diff
@@ -0,0 +1,122 @@
...
+ 	lambda_deg	= gpsOut.y; 		//degrees
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. `lambda_deg = gps_out.y;`

File: gps_rviz/src/gps_handler.cpp [hunk 29]
```diff
@@ -0,0 +1,122 @@
...
+	phi_rad 	= phi_deg/180*PI; 	//radians
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. `phi_rad = phi_deg * M_PI / 180.0;`

File: gps_rviz/src/gps_handler.cpp [hunk 30]
```diff
@@ -0,0 +1,122 @@
...
+	lambda_rad 	= lambda_deg/180*PI; 	//radians
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. `lambda_rad = lambda_deg * M_PI / 180.0;`

File: gps_rviz/src/gps_handler.cpp [hunk 31]
```diff
@@ -0,0 +1,122 @@
...
+
+	geometry_msgs::Vector3 gps_cart;
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. `gps_cartesian`

File: gps_rviz/src/gps_handler.cpp [hunk 32]
```diff
@@ -0,0 +1,122 @@
...
+	gps_cart.x = (R+height)*cos(phi_rad)*cos(lambda_rad); // WGS84
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. 

```
/// Using WGS84 convention.
gps_cartesian.x = (r_meter + height_meter) * cos(phi_rad) * cos(lambda_rad);
```

File: gps_rviz/src/gps_handler.cpp [hunk 33]
```diff
@@ -0,0 +1,122 @@
...
+	gps_cart.y = (R+height)*cos(phi_rad)*sin(lambda_rad);
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. 
`gps_cartesian.y = (r_meter + height_meter) * cos(phi_rad) * sin(lambda_rad);`

File: gps_rviz/src/gps_handler.cpp [hunk 34]
```diff
@@ -0,0 +1,122 @@
...
+	gps_cart.z = (R*(1-e*e)+height)*sin(phi_rad);
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE.
`gps_cartesian.z = (r_meter * (1.0 - e * e) + height_meter) * sin(phi_rad);`

File: gps_rviz/src/gps_handler.cpp [hunk 35]
```diff
@@ -0,0 +1,122 @@
...
+
+	ROS_INFO("In cartesian coordinates this is: [%f]||[%f]||[%f]", gps_cart.x, gps_cart.y,gps_cart.z);
+return gps_cart;
+}
+
+// ----------------------------------------------
+// Add a single point (x,y,z) to MarkerArray GPS_TRACK
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. Sentensify.

File: gps_rviz/src/gps_handler.cpp [hunk 36]
```diff
@@ -0,0 +1,122 @@
...
+	gps_cart.y = (R+height)*cos(phi_rad)*sin(lambda_rad);
+	gps_cart.z = (R*(1-e*e)+height)*sin(phi_rad);
+
+	ROS_INFO("In cartesian coordinates this is: [%f]||[%f]||[%f]", gps_cart.x, gps_cart.y,gps_cart.z);
+return gps_cart;
+}
+
+// ----------------------------------------------
+// Add a single point (x,y,z) to MarkerArray GPS_TRACK 
+void addToTrack(const geometry_msgs::Vector3 PointToBeAddedToTrack){
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. 
`const geometry_msgs::Vector3& point_to_be_added_to_track`
const ref or pointer

File: gps_rviz/src/gps_handler.cpp [hunk 37]
```diff
@@ -0,0 +1,122 @@
...
+
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. Remove vspace.

File: gps_rviz/src/gps_handler.cpp [hunk 38]
```diff
@@ -0,0 +1,122 @@
...
+	geometry_msgs::Point p;
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. Try to use good names `point`.

File: gps_rviz/src/gps_handler.cpp [hunk 39]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+// ------------------------------------------------------------
+// Define parameters of GPS_TRACK
+void initializeTrack(){
+	gps_points.header.frame_id = gps_line_strip.header.frame_id = gps_line_list.header.frame_id = "map"; 
+	gps_points.header.stamp = gps_line_strip.header.stamp = gps_line_list.header.stamp = ros::Time::now();
+	gps_points.ns = gps_line_strip.ns =gps_line_list.ns = "gps_points_and_lines";
+	gps_points.action = gps_line_strip.action = gps_line_list.action = visualization_msgs::Marker::ADD;
+	gps_points.pose.orientation.w = gps_line_strip.pose.orientation.w = gps_line_list.pose.orientation.w = 1.0;
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: Don't use more than 100 characters in a line (100C convection)

File: gps_rviz/src/gps_handler.cpp [hunk 40]
```diff
@@ -0,0 +1,122 @@
...
+	gps_line_list.points.push_back(p);
+
+	GPS_TRACK.markers.push_back(gps_points);
+	GPS_TRACK.markers.push_back(gps_line_strip);
+	GPS_TRACK.markers.push_back(gps_line_list);
+	ROS_INFO("Adding point [%f],[%f], [%f] in mission frame to Track.", PointToBeAddedToTrack.x,PointToBeAddedToTrack.y,PointToBeAddedToTrack.z); 	
+}
+
+// ------------------------------------------------------------
+// Define parameters of GPS_TRACK
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. Sentensify.

File: gps_rviz/src/gps_handler.cpp [hunk 41]
```diff
@@ -0,0 +1,122 @@
...
+void initializeTrack(){
+	gps_points.header.frame_id = gps_line_strip.header.frame_id = gps_line_list.header.frame_id = "map";
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. 2pt indentation.

File: gps_rviz/src/gps_handler.cpp [hunk 42]
```diff
@@ -0,0 +1,122 @@
...
+
+	GPS_TRACK.markers.push_back(gps_points);
+	GPS_TRACK.markers.push_back(gps_line_strip);
+	GPS_TRACK.markers.push_back(gps_line_list);
+	ROS_INFO("Adding point [%f],[%f], [%f] in mission frame to Track.", PointToBeAddedToTrack.x,PointToBeAddedToTrack.y,PointToBeAddedToTrack.z); 	
+}
+
+// ------------------------------------------------------------
+// Define parameters of GPS_TRACK
+void initializeTrack(){
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. `void initializeTrack() {`

File: gps_rviz/src/gps_handler.cpp [hunk 43]
```diff
@@ -0,0 +1,122 @@
...
+	gps_line_strip.color.b = 1.0;
+	gps_line_strip.color.a = 1.0;	
+	gps_line_list.color.r = 1.0;
+	gps_line_list.color.a = 1.0;
+
+	ROS_INFO("Defined type, scale and color successfully");
+}
+
+// ----------------------------------------------------------------------
+} //end namespace gps
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: `} // namespace gps`

File: gps_rviz/src/gps_handler.h [hunk 44]
```diff
@@ -0,0 +1,14 @@
+#ifndef GPS_HANDLER_H_
+#define GPS_HANDLER_H_
+
+#include "geometry_msgs/Vector3.h"
+
+const float InitializingPoint_x =4266066; //TODO Define this point
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.h]: DONE. 
For constants we write
`const float kInitializingPointX = 4266066.0;`

File: gps_rviz/src/gps_handler.h [hunk 45]
```diff
@@ -0,0 +1,14 @@
...
+const float InitializingPoint_y = 584902;
+const float InitializingPoint_z =4691134; 
+
+const float majorAxis = 6378137; //in meters
+const float minorAxis 	= 6356752.31; //in meters
+
+
```
Comment (hitimo) [inline gps_rviz/src/gps_handler.h:10]: DONE. Remove vspace.

File: gps_rviz/src/gps_rviz.cpp [hunk 46]
```diff
@@ -1,152 +1,63 @@
 #include "ros/ros.h"
 #include "sensor_msgs/NavSatFix.h"
 #include "geometry_msgs/Vector3.h"
-#include "visualization_msgs/Marker.h"
-#include "math.h"
-
+#include "gps_handler.cpp"
+#include "gps_rviz.h"
```
Comment (hitimo) [inline gps_rviz/src/gps_rviz.cpp:1]: DONE. 
We use this include style: https://github.com/ethz-asl/multiagent_mapping/wiki/Header-include-guide
Alphabetize includes.

File: gps_rviz/src/gps_rviz.cpp [hunk 47]
```diff
@@ -1,152 +1,63 @@
...
-	geometry_msgs::Vector3 gps_cart;
-	gps_cart.x = (R+height)*cos(phi_rad)*cos(lambda_rad); //TODO introduce WGS84!!
-	gps_cart.y = (R+height)*cos(phi_rad)*sin(lambda_rad);
-	gps_cart.z = (R*(1-e*e)+height)*sin(phi_rad);
-
-	ROS_INFO("In cartesian coordinates this is: [%f]||[%f]||[%f]", gps_cart.x, gps_cart.y,gps_cart.z);
-return gps_cart;
-}
 
+void publishGPS_Track(){
```
Comment (hitimo) [inline gps_rviz/src/gps_rviz.cpp]: DONE. `void publishGpsTrack() {`

File: gps_rviz/src/gps_rviz.cpp [hunk 48]
```diff
@@ -1,152 +1,63 @@
...
-		gps_pub.publish(gps_line_strip);
-		gps_pub.publish(gps_line_list);
-	
-		ROS_INFO("Published successfully.");
-
-		ros::spin();
-
-	} //end while ros::ok()
-	
+	ros::NodeHandle node_handle_;
```
Comment (hitimo) [inline gps_rviz/src/gps_rviz.cpp]: DONE. 2pt-indentation.

Comment (hitimo) [inline gps_rviz/src/gps_rviz.cpp]: DONE. `ros::NodeHandle node_handle;`
Underscore only for class members.

File: gps_rviz/src/gps_rviz.cpp [hunk 49]
```diff
@@ -1,152 +1,63 @@
...
-		gps_pub.publish(gps_line_list);
-	
-		ROS_INFO("Published successfully.");
-
-		ros::spin();
-
-	} //end while ros::ok()
-	
+	ros::NodeHandle node_handle_; 
+	ros::Publisher gps_pub = node_handle_.advertise<visualization_msgs::MarkerArray>("GPS_marker",1000);
```
Comment (hitimo) [inline gps_rviz/src/gps_rviz.cpp]: DONE. Horizontal space after `,`

File: gps_rviz/src/gps_rviz.cpp [hunk 50]
```diff
@@ -1,152 +1,63 @@
...
+	gps_pub.publish(gps::GPS_TRACK);
+	ros::spin();
 }
 
-void gps2rviz(const sensor_msgs::NavSatFix gps_NavSatFix){ //callback function
+void gps_callback(const sensor_msgs::NavSatFix gps_NavSatFix){ //callback function
```
Comment (hitimo) [inline gps_rviz/src/gps_rviz.cpp]: DONE. Comment seems superfluous given the function name.

Comment (hitimo) [inline gps_rviz/src/gps_rviz.cpp]: DONE. gpsCallback

Comment (hitimo) [inline gps_rviz/src/gps_rviz.cpp]: DONE. `const sensor_msgs::NavSatFix&`

File: gps_rviz/src/gps_rviz.cpp [hunk 51]
```diff
@@ -1,152 +1,63 @@
...
 
 	geometry_msgs::Vector3 gpsOut;
 
-	gpsOut.x = gps_NavSatFix.latitude; 
-	gpsOut.y = gps_NavSatFix.longitude;
+	gpsOut.x = gps_NavSatFix.latitude; // TODO this is sort of a hack...
```
Comment (hitimo) [inline gps_rviz/src/gps_rviz.cpp]: DONE. 
We use this todo style:
`TODO(andschaf): Your comment here.`

Comment (hitimo) [inline gps_rviz/src/gps_rviz.cpp]: DONE. `gps_out`

File: gps_rviz/src/gps_rviz.cpp [hunk 52]
```diff
@@ -1,152 +1,63 @@
...
-
-	ROS_INFO("I extracted: lat=[%f] , lon=[%f] and alt=[%f]" ,gpsOut.x, gpsOut.y, gpsOut.z);
 	
-	geometry_msgs::Vector3 cleaned_GPS_point;
-	cleaned_GPS_point = gps2cart(gpsOut);	
+	geometry_msgs::Vector3 GPS_point_WGS84;
+	GPS_point_WGS84 = gps::gps2cart(gpsOut);	
+
+	if (!gps::REFERENCE_FRAME_IS_INITIALIZED){
+	 ROS_INFO("Initializing Reference Frame...");
```
Comment (hitimo) [inline gps_rviz/src/gps_rviz.cpp]: DONE. 2pt-indentation.

File: gps_rviz/src/gps_rviz.cpp [hunk 53]
```diff
@@ -1,152 +1,63 @@
...
+	 gps::InitializingPoint = GPS_point_WGS84;
+	 gps::alpha_init = PI + gps_NavSatFix.longitude/180*PI;
```
Comment (hitimo) [inline gps_rviz/src/gps_rviz.cpp]: `M_PI`

Comment (hitimo) [inline gps_rviz/src/gps_rviz.cpp]: `/ 180.0 * M_PI;`

File: gps_rviz/src/gps_rviz.cpp [hunk 54]
```diff
@@ -1,152 +1,63 @@
...
+	 gps::beta_init = PI/2 - gps_NavSatFix.latitude/180*PI;
```
Comment (hitimo) [inline gps_rviz/src/gps_rviz.cpp]: `M_PI / 2.0 - gps_NavSatFix.latitude / 180.0 * M_PI;`

File: gps_rviz/src/gps_rviz.cpp [hunk 55]
```diff
@@ -1,152 +1,63 @@
...
+	 gps::REFERENCE_FRAME_IS_INITIALIZED = true;
 	
-	ROS_INFO("in gps2rviz I return [%f]||[%f]||[%f]", cleaned_GPS_point.x, cleaned_GPS_point.y, cleaned_GPS_point.z);
-	publishGPSpositions(cleaned_GPS_point);
+	 gps::InitializingPoint_missionframe = gps::transformation(gps::InitializingPoint, gps::alpha_init, gps::beta_init);
+
```
Comment (hitimo) [inline gps_rviz/src/gps_rviz.cpp:40]: Remove vspace.

File: gps_rviz/src/gps_handler.cpp (hunk 8 above)
Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: DONE. What is `a`, `b`? Better names.

**Comment by hitimo** [comment]
Thanks @andschaf 
Some style comments, otherwise this looks good to me.

File: gps_rviz/src/gps_handler.cpp (hunk 2 above)
Comment (andschaf) [inline gps_rviz/src/gps_handler.cpp]: Done

File: gps_rviz/src/gps_handler.cpp (hunk 39 above)
Comment (andschaf) [inline gps_rviz/src/gps_handler.cpp]: is there a way to do line breaks?

Comment (hitimo) [inline gps_rviz/src/gps_handler.cpp]: You can simply have a line break after the equal sign.

**Pull request merged**